contracts/SpokePool.sol

```diff
@@ -147,19 +182,73 @@ abstract contract SpokePool is Testable, Lockable, MultiCaller {
     }
 
     function initiateRelay(
-        uint256 originChain,
-        address sender,
+        uint256 originChainId,
         uint256 amount,
-        address recipient,
-        uint256 relayerFee,
-        uint256 realizedLpFee
-    ) public {}
+        uint64 depositId,
+        uint64 relayerFeePct,
+        uint64 realizedLpFeePct,
+        address destinationToken,
+        address sender,
+        address recipient
+    ) public {
+        // We limit the relay fees to prevent the user spending all their funds on fees.
+        require(relayerFeePct <= 0.5e18 && realizedLpFeePct <= 0.5e18, "invalid fees");
+
+        // Use new relay count as unique ID for new relay.
+        relays.push(
+            RelayData(
+                recipient,
+                destinationToken,
+                relayerFeePct,
+                realizedLpFeePct,
+                amount, // total relay amount
+                0 // total amount filled
+            )
+        );
+        uint64 relayId = uint64(relays.length - 1);
+
+        emit InitiatedRelay(
+            originChainId,
+            amount,
+            depositId,
+            relayId,
+            relayerFeePct,
+            realizedLpFeePct,
+            destinationToken,
+            sender,
+            recipient,
+            msg.sender
+        );
+    }
 
     function fillRelay(
-        uint256 relayId,
-        uint256 fillAmount,
+        uint64 relayId,
+        uint256 amount,
         uint256 repaymentChain
-    ) public {}
+    ) public {
+        RelayData memory relay = relays[relayId];
+        // The following check will fail if:
+        // - relay has not been instantiated and relayAmount = 0.
+        // - caller's desired amount to fill would send filledAmount over relayAmount.
+        require(amount > 0 && relay.relayAmount >= relay.filledAmount + amount, "Invalid remaining relay amount");
+
+        // Update total filled amount with this new fill. Each fill can be uniquely identified by the
+        // total filled amount including itself, the relay ID, and the chain ID of this contract.
+        relays[relayId].filledAmount += amount;
+
+        // Pull fill amount net fees from caller, which is the amount owed to the recipient. The relayer will receive
+        // this amount plus the relayer fee after the relayer refund is processed.
+        uint256 amountNetFees = amount - _getAmountFromPct(relay.realizedLpFeePct + relay.relayerFeePct, amount);
+        IERC20(relay.destinationToken).safeTransferFrom(msg.sender, relay.recipient, amountNetFees);
+
+        // If relay token is weth then unwrap and send eth.
```

Comment:
do you think that we should try let the user choose to receive WETH or ETH? it's kinda strange that a user sending WETH or ETH always gets ETH? (obs not something for now but I'm curious what you think).

Reply:
the depositor would need an extra bool, which is kind of a pain 🤔 . might not be worth it tbh